engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java

```diff
@@ -52,4 +69,174 @@ public void setTypeName(QualifiedName typeName) {
    */
   @Override
   public abstract String getName();
+
+  protected static final class ArgContext {
+    private TruffleObject returnValue;
+    private EnsoHashMap warnings;
+
+    public ArgContext() {}
+
+    public TruffleObject getReturnValue() {
+      return returnValue;
+    }
+
+    public boolean hasWarnings() {
+      return this.warnings != null;
+    }
+
+    private void addWarnings(
+        VirtualFrame frame, HashMapInsertAllNode insertNode, EnsoHashMap newWarnings) {
+      if (this.warnings == null) {
+        this.warnings = newWarnings;
+      } else {
+        int maxWarnings = EnsoContext.get(insertNode).getWarningsLimit();
+        this.warnings = insertNode.executeInsertAll(frame, this.warnings, newWarnings, maxWarnings);
+      }
+    }
+  }
+
+  protected abstract static class ArgNode extends Node {
+    private static final byte IS_SELF = 0x01;
+    private static final byte IS_ARRAY = 0x02;
+    private static final byte REQUIRES_CAST = 0x04;
+    private static final byte CHECK_ERRORS = 0x08;
+    private static final byte CHECK_PANIC_SENTINEL = 0x10;
+    private static final byte CHECK_WARNINGS = 0x20;
+    private final byte flags;
+    @CompilerDirectives.CompilationFinal private Type ensoType;
+
+    @Child private WarningsLibrary warnings;
+    @Child private AppendWarningNode appendWarningNode;
+    @Child private HashMapInsertAllNode mapInsertAllNode;
+
+    ArgNode(byte flags) {
+      this.flags = flags;
+      if (is(CHECK_WARNINGS)) {
+        this.warnings = WarningsLibrary.getFactory().createDispatched(5);
+      }
+    }
+
+    private boolean is(byte what) {
+      return (flags & what) != 0;
+    }
+
+    @SuppressWarnings("unchecked")
+    public final <T> T processArgument(
+        VirtualFrame frame, Class<T> type, Object value, ArgContext context) {
+      assert value != null;
+      if (is(CHECK_ERRORS) && value instanceof DataflowError err) {
+        context.returnValue = err;
+        return null;
+      }
+      if (is(CHECK_PANIC_SENTINEL) && value instanceof PanicSentinel sentinel) {
+        throw sentinel.getPanic();
+      }
+      if (warnings != null) {
+        if (warnings.hasWarnings(value)) {
+          if (mapInsertAllNode == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            this.mapInsertAllNode = HashMapInsertAllNode.build();
+          }
+          if (this.mapInsertAllNode != null) {
+            try {
+              context.addWarnings(frame, mapInsertAllNode, warnings.getWarnings(value, false));
+              value = warnings.removeWarnings(value);
+            } catch (UnsupportedMessageException ex) {
+              throw raise(RuntimeException.class, ex);
+            }
+          }
+        }
+      }
+      if (is(REQUIRES_CAST)) {
+        var ctx = EnsoContext.get(this);
+        if (this.ensoType == null) {
```

Comment:
The first time `ArgNode` is requested to "cast", it searches for the `ensoType` representing the value of the argument. It would be better to do this when the `ArgNode` is being created, but it may be too early (before `Standard.Base` types are loaded in), so I haven't even tried.

Opinions?